File: backend/api/app/domain/use_cases/post/update.py
from core.use_cases.use_case import UseCase
from app.domain.repositories.post import BaseRepository
from core.common.equatable import Equatable
from core.common.either import Either
from core.errors.failure import Failure
from app.domain.entities.post import Post, PostEntity

class Params(Equatable):
    def __init__(self, post: Post, post_id: str) -> None:
        self.post = post
        self.post_id = post_id
        
class UpdatePost(UseCase[PostEntity]):
    def __init__(self, repository: BaseRepository):
        self.repository = repository
    
    async def __call__(self, params: Params) -> Either[Failure, PostEntity]:
        return await self.repository.update_post(params.post, params.post_id)

File: mobile/lib/features/architect/domains/use_cases/post/update.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';

import '../../../../../core/errors/failure.dart';
import '../../../../../core/use_cases/usecase.dart';
import '../../entities/post.dart';
import '../../repositories/post.dart';

class UpdatePost implements UseCase<Post, Params> {
  final PostRepository repository;

  const UpdatePost(this.repository);

  @override
  Future<Either<Failure, Post>> call(Params params) async {
    return await repository.update(
      id: params.id,
      image: params.image,
      title: params.title,
      content: params.content,
      userId: params.userId,
      tags: params.tags,
    );
  }
}

class Params extends Equatable {
  final String id;
  final String image;
  final String title;
  final String? content;
  final String userId;
  final List<String> tags;

  const Params({
    required this.id,
    required this.image,
    required this.title,
    this.content,
    required this.userId,
    required this.tags,
  });

  @override
  List<Object?> get props => [image, title, content, userId, tags];
}
